lxd/identities.go

```diff
@@ -277,137 +350,198 @@ func identityAccessHandler(entitlement auth.Entitlement) func(d *Daemon, r *http
+				return canViewIdentity(entity.IdentityURL(string(id.AuthMethod), id.Identifier))
+			}
 
-	var identities []dbCluster.Identity
-	var groupsByIdentityID map[int][]dbCluster.AuthGroup
-	var apiIdentity *api.Identity
-	err = s.DB.Cluster.Transaction(r.Context(), func(ctx context.Context, tx *db.ClusterTx) error {
-		// Get all identities, filter by authentication method if present.
-		var filters []dbCluster.IdentityFilter
-		if authenticationMethod != "" {
-			clusterAuthMethod := dbCluster.AuthMethod(authenticationMethod)
-			filters = append(filters, dbCluster.IdentityFilter{AuthMethod: &clusterAuthMethod})
+			return canViewCertificate(entity.CertificateURL(id.Identifier))
 		}
 
-		allIdentities, err := dbCluster.GetIdentitys(ctx, tx.Tx(), filters...)
+		canViewGroup, err := s.Authorizer.GetPermissionChecker(r.Context(), auth.EntitlementCanView, entity.TypeAuthGroup)
 		if err != nil {
-			return err
+			return response.SmartError(err)
 		}
 
-		// Filter results by what the user is allowed to view.
-		for _, id := range allIdentities {
-			if canView(id) {
-				identities = append(identities, id)
+		var identities []dbCluster.Identity
+		var groupsByIdentityID map[int][]dbCluster.AuthGroup
+		var apiIdentity *api.Identity
+		err = s.DB.Cluster.Transaction(r.Context(), func(ctx context.Context, tx *db.ClusterTx) error {
+			// Get all identities, filter by authentication method if present.
+			var filters []dbCluster.IdentityFilter
+			if authenticationMethod != "" {
+				clusterAuthMethod := dbCluster.AuthMethod(authenticationMethod)
+				filters = append(filters, dbCluster.IdentityFilter{AuthMethod: &clusterAuthMethod})
 			}
-		}
-
-		if len(identities) == 0 {
-			return nil
-		}
 
-		if recursion == "1" && len(identities) == 1 {
-			// It's likely that the user can only view themselves. If so we can optimise here by only getting the
-			// groups for that user.
-			apiIdentity, err = identities[0].ToAPI(ctx, tx.Tx(), canViewGroup)
+			allIdentities, err := dbCluster.GetIdentitys(ctx, tx.Tx(), filters...)
 			if err != nil {
 				return err
 			}
-		} else if recursion == "1" {
-			// Otherwise, get all groups and populate the identities outside of the transaction.
-			groupsByIdentityID, err = dbCluster.GetAllAuthGroupsByIdentityIDs(ctx, tx.Tx())
-			if err != nil {
-				return err
+
+			// Filter results by what the user is allowed to view.
+			for _, id := range allIdentities {
+				if canView(id) {
+					identities = append(identities, id)
+				}
 			}
-		}
 
-		return nil
-	})
-	if err != nil {
-		return response.SmartError(err)
-	}
+			if len(identities) == 0 {
+				return nil
+			}
 
-	// Optimisation for user that can only view themselves.
-	if apiIdentity != nil {
-		return response.SyncResponse(true, []api.Identity{*apiIdentity})
-	}
+			if recursion == "1" && len(identities) == 1 {
+				// It's likely that the user can only view themselves. If so we can optimise here by only getting the
```

Comment:
can we check the contents of identities[0] to be more confident than "likely"?

Reply:
The optimisation is for when the caller can only see one identity. It doesn't matter if it is the caller. E.g. the caller may be using the unix socket and there is only one identity in the system. I'll rephrase the comments to make it clear.

Reply:
Done.